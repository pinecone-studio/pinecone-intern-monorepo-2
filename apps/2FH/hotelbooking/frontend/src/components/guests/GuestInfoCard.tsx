'use client';
import React from 'react';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import CheckoutModal from './CheckoutModal';

interface GuestInfoCardProps {
  guestInfo: {
    firstName: string;
    lastName: string;
    email: string;
    phone: string;
    guestRequest: string;
    roomNumber: string;
  };
  booking: {
    id: string;
    status: string;
    adults?: number | null;
    children?: number | null;
    checkInDate: any;
    checkOutDate: any;
  };
  onStatusUpdate: () => void;
}

const GuestInfoCard = ({ guestInfo, booking, onStatusUpdate }: GuestInfoCardProps) => {
  const formatDate = (date: any) => {
    const d = new Date(date);
    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    return {
      day: days[d.getDay()],
      month: months[d.getMonth()],
      date: d.getDate(),
      time: d.toLocaleTimeString('en-US', { hour: 'numeric', minute: '2-digit', hour12: true }),
    };
  };

  const getStatusBadgeStyle = (status: string) => {
    switch (status) {
      case 'BOOKED':
        return 'bg-blue-600 text-white';
      case 'COMPLETED':
        return 'bg-green-600 text-white';
      case 'CANCELLED':
        return 'bg-orange-500 text-white';
      default:
        return 'bg-blue-600 text-white';
    }
  };

  const checkInFormatted = formatDate(booking.checkInDate);
  const checkOutFormatted = formatDate(booking.checkOutDate);

  return (
    <Card>
      <CardHeader>
        <CardTitle className="text-lg font-semibold">Guest Info</CardTitle>
      </CardHeader>
      <CardContent>
        <div className="grid grid-cols-2 gap-4">
          <div>
            <label className="text-sm font-medium text-gray-600">First Name</label>
            <p className="text-gray-900">{guestInfo.firstName}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Last Name</label>
            <p className="text-gray-900">{guestInfo.lastName}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Status</label>
            <div className="mt-1">
              <Badge className={getStatusBadgeStyle(booking.status)}>{booking.status.charAt(0) + booking.status.slice(1).toLowerCase()}</Badge>
            </div>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Guests</label>
            <p className="text-gray-900">
              {(booking as any).adults || 1} adult{((booking as any).adults || 1) > 1 ? 's' : ''}, {(booking as any).children || 0} children
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Check in</label>
            <p className="text-gray-900">
              {checkInFormatted.month} {checkInFormatted.date}, {checkInFormatted.day}, {checkInFormatted.time}
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Check out</label>
            <p className="text-gray-900">
              {checkOutFormatted.month} {checkOutFormatted.date}, {checkOutFormatted.day}, {checkOutFormatted.time}
            </p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Email</label>
            <p className="text-gray-900">{guestInfo.email}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Phone</label>
            <p className="text-gray-900">{guestInfo.phone}</p>
          </div>
          <div>
            <label className="text-sm font-medium text-gray-600">Guest Request</label>
            <p className="text-gray-900">{guestInfo.guestRequest}</p>
          </div>
        </div>
        <div className="flex justify-end mt-6">
          <CheckoutModal bookingId={booking.id} currentStatus={booking.status} onStatusUpdate={onStatusUpdate} />
        </div>
      </CardContent>
    </Card>
  );
};

export default GuestInfoCard;
