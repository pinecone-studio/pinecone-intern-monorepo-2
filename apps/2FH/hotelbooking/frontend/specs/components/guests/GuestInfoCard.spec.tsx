/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import GuestInfoCard from '@/components/guests/GuestInfoCard';

// Mock dependencies
jest.mock('@/components/guests/CheckoutModal', () => ({
  __esModule: true,
  default: ({ bookingId, currentStatus, onStatusUpdate }: any) => (
    <div data-testid="checkout-modal">
      <div>Checkout Modal - {currentStatus}</div>
      <button onClick={onStatusUpdate}>Update Status</button>
    </div>
  ),
}));

describe('GuestInfoCard', () => {
  const mockGuestInfo = {
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@example.com',
    phone: '+976 99112233',
    guestRequest: 'Late check-in requested',
    roomNumber: 'Room #101',
  };

  const mockBooking = {
    id: 'booking-1',
    status: 'BOOKED',
    adults: 2,
    children: 1,
    checkInDate: '2024-02-01T14:00:00Z',
    checkOutDate: '2024-02-03T11:00:00Z',
  };

  const defaultProps = {
    guestInfo: mockGuestInfo,
    booking: mockBooking,
    onStatusUpdate: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders guest information correctly', () => {
    render(<GuestInfoCard {...defaultProps} />);

    expect(screen.getByText('Guest Info')).toBeInTheDocument();
    expect(screen.getByText('John')).toBeInTheDocument();
    expect(screen.getByText('Doe')).toBeInTheDocument();
    expect(screen.getByText('john.doe@example.com')).toBeInTheDocument();
    expect(screen.getByText('+976 99112233')).toBeInTheDocument();
    expect(screen.getByText('Late check-in requested')).toBeInTheDocument();
  });

  it('displays booking status with correct styling', () => {
    render(<GuestInfoCard {...defaultProps} />);

    const statusBadge = screen.getByText('Booked');
    expect(statusBadge).toHaveClass('bg-blue-600', 'text-white');
  });

  it('displays guest count correctly', () => {
    render(<GuestInfoCard {...defaultProps} />);

    expect(screen.getByText('2 adults, 1 children')).toBeInTheDocument();
  });

  it('displays guest count with singular forms', () => {
    const bookingWithSingular = {
      ...mockBooking,
      adults: 1,
      children: 0,
    };

    render(<GuestInfoCard {...defaultProps} booking={bookingWithSingular} />);

    expect(screen.getByText('1 adult, 0 children')).toBeInTheDocument();
  });

  it('formats check-in date correctly', () => {
    render(<GuestInfoCard {...defaultProps} />);

    // The date should be formatted as "Feb 1, Thursday, 10:00 PM"
    expect(screen.getByText(/Feb 1, Thursday, 10:00 PM/)).toBeInTheDocument();
  });

  it('formats check-out date correctly', () => {
    render(<GuestInfoCard {...defaultProps} />);

    // The date should be formatted as "Feb 3, Saturday, 7:00 PM"
    expect(screen.getByText(/Feb 3, Saturday, 7:00 PM/)).toBeInTheDocument();
  });

  it('handles different booking statuses', () => {
    const completedBooking = { ...mockBooking, status: 'COMPLETED' };
    render(<GuestInfoCard {...defaultProps} booking={completedBooking} />);

    const statusBadge = screen.getByText('Completed');
    expect(statusBadge).toHaveClass('bg-green-600', 'text-white');
  });

  it('handles cancelled booking status', () => {
    const cancelledBooking = { ...mockBooking, status: 'CANCELLED' };
    render(<GuestInfoCard {...defaultProps} booking={cancelledBooking} />);

    const statusBadge = screen.getByText('Cancelled');
    expect(statusBadge).toHaveClass('bg-orange-500', 'text-white');
  });

  it('handles unknown booking status', () => {
    const unknownBooking = { ...mockBooking, status: 'UNKNOWN' };
    render(<GuestInfoCard {...defaultProps} booking={unknownBooking} />);

    const statusBadge = screen.getByText('Unknown');
    expect(statusBadge).toHaveClass('bg-blue-600', 'text-white');
  });

  it('renders checkout modal', () => {
    render(<GuestInfoCard {...defaultProps} />);

    expect(screen.getByTestId('checkout-modal')).toBeInTheDocument();
    expect(screen.getByText('Checkout Modal - BOOKED')).toBeInTheDocument();
  });

  it('handles status update', () => {
    const onStatusUpdate = jest.fn();
    render(<GuestInfoCard {...defaultProps} onStatusUpdate={onStatusUpdate} />);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    expect(onStatusUpdate).toHaveBeenCalled();
  });

  it('handles missing guest information', () => {
    const incompleteGuestInfo = {
      firstName: 'John',
      lastName: '',
      email: '',
      phone: '',
      guestRequest: '',
      roomNumber: '',
    };

    render(<GuestInfoCard {...defaultProps} guestInfo={incompleteGuestInfo} />);

    expect(screen.getByText('John')).toBeInTheDocument();
    // Check that lastName field exists but is empty
    const lastNameField = screen.getByText('Last Name').closest('div')?.querySelector('p');
    expect(lastNameField).toBeInTheDocument();
    expect(lastNameField?.textContent).toBe('');
  });

  it('handles null/undefined guest information', () => {
    const nullGuestInfo = {
      firstName: null,
      lastName: undefined,
      email: null,
      phone: undefined,
      guestRequest: null,
      roomNumber: undefined,
    };

    render(<GuestInfoCard {...defaultProps} guestInfo={nullGuestInfo} />);

    // Should render without crashing
    expect(screen.getByText('Guest Info')).toBeInTheDocument();
  });

  it('handles missing booking information', () => {
    const incompleteBooking = {
      id: 'booking-1',
      status: 'BOOKED',
      adults: null,
      children: undefined,
      checkInDate: null,
      checkOutDate: undefined,
    };

    render(<GuestInfoCard {...defaultProps} booking={incompleteBooking} />);

    // Should handle null/undefined values gracefully
    expect(screen.getByText('Guest Info')).toBeInTheDocument();
  });

  it('renders with proper grid layout', () => {
    render(<GuestInfoCard {...defaultProps} />);

    const gridContainer = document.querySelector('.grid.grid-cols-2.gap-4');
    expect(gridContainer).toBeInTheDocument();
  });

  it('renders checkout modal at the bottom', () => {
    render(<GuestInfoCard {...defaultProps} />);

    const checkoutModal = screen.getByTestId('checkout-modal');
    const parentContainer = checkoutModal.closest('.flex.justify-end.mt-6');
    expect(parentContainer).toBeInTheDocument();
  });

  it('handles different date formats', () => {
    const bookingWithDifferentDates = {
      ...mockBooking,
      checkInDate: '2024-12-25T00:00:00Z',
      checkOutDate: '2024-12-27T23:59:59Z',
    };

    render(<GuestInfoCard {...defaultProps} booking={bookingWithDifferentDates} />);

    // Should format dates correctly regardless of format
    expect(screen.getByText(/Dec 25/)).toBeInTheDocument();
    expect(screen.getByText(/Dec 28/)).toBeInTheDocument();
  });

  it('handles edge case dates', () => {
    const bookingWithEdgeDates = {
      ...mockBooking,
      checkInDate: '2024-01-01T00:00:00Z',
      checkOutDate: '2024-12-31T23:59:59Z',
    };

    render(<GuestInfoCard {...defaultProps} booking={bookingWithEdgeDates} />);

    const jan1Elements = screen.getAllByText(/Jan 1/);
    expect(jan1Elements.length).toBeGreaterThanOrEqual(1);
    // Both check-in and check-out are Jan 1 in this test case
    expect(jan1Elements.length).toBeGreaterThanOrEqual(2);
  });

  it('handles invalid dates gracefully', () => {
    const bookingWithInvalidDates = {
      ...mockBooking,
      checkInDate: 'invalid-date',
      checkOutDate: 'another-invalid-date',
    };

    render(<GuestInfoCard {...defaultProps} booking={bookingWithInvalidDates} />);

    // Should not crash and should render the component
    expect(screen.getByText('Guest Info')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<GuestInfoCard {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with different status', () => {
    const completedBooking = { ...mockBooking, status: 'COMPLETED' };
    const { container } = render(<GuestInfoCard {...defaultProps} booking={completedBooking} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
