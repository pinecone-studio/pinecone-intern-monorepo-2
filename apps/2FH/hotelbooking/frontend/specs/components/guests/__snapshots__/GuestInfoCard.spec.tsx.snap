// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestInfoCard matches snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="tracking-tight text-lg font-semibold"
    >
      Guest Info
    </h3>
  </div>
  <div
    class="p-6 pt-0"
  >
    <div
      class="grid grid-cols-2 gap-4"
    >
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          First Name
        </label>
        <p
          class="text-gray-900"
        >
          John
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Last Name
        </label>
        <p
          class="text-gray-900"
        >
          Doe
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Status
        </label>
        <div
          class="mt-1"
        >
          <div
            class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent hover:bg-primary/80 bg-blue-600 text-white"
          >
            Booked
          </div>
        </div>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Guests
        </label>
        <p
          class="text-gray-900"
        >
          2
           adult
          s
          , 
          1
           children
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Check in
        </label>
        <p
          class="text-gray-900"
        >
          Feb
           
          1
          , 
          Thursday
          , 
          10:00 PM
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Check out
        </label>
        <p
          class="text-gray-900"
        >
          Feb
           
          3
          , 
          Saturday
          , 
          7:00 PM
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Email
        </label>
        <p
          class="text-gray-900"
        >
          john.doe@example.com
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Phone
        </label>
        <p
          class="text-gray-900"
        >
          +976 99112233
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Guest Request
        </label>
        <p
          class="text-gray-900"
        >
          Late check-in requested
        </p>
      </div>
    </div>
    <div
      class="flex justify-end mt-6"
    >
      <div
        data-testid="checkout-modal"
      >
        <div>
          Checkout Modal - 
          BOOKED
        </div>
        <button>
          Update Status
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`GuestInfoCard matches snapshot with different status 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="tracking-tight text-lg font-semibold"
    >
      Guest Info
    </h3>
  </div>
  <div
    class="p-6 pt-0"
  >
    <div
      class="grid grid-cols-2 gap-4"
    >
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          First Name
        </label>
        <p
          class="text-gray-900"
        >
          John
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Last Name
        </label>
        <p
          class="text-gray-900"
        >
          Doe
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Status
        </label>
        <div
          class="mt-1"
        >
          <div
            class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent hover:bg-primary/80 bg-green-600 text-white"
          >
            Completed
          </div>
        </div>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Guests
        </label>
        <p
          class="text-gray-900"
        >
          2
           adult
          s
          , 
          1
           children
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Check in
        </label>
        <p
          class="text-gray-900"
        >
          Feb
           
          1
          , 
          Thursday
          , 
          10:00 PM
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Check out
        </label>
        <p
          class="text-gray-900"
        >
          Feb
           
          3
          , 
          Saturday
          , 
          7:00 PM
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Email
        </label>
        <p
          class="text-gray-900"
        >
          john.doe@example.com
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Phone
        </label>
        <p
          class="text-gray-900"
        >
          +976 99112233
        </p>
      </div>
      <div>
        <label
          class="text-sm font-medium text-gray-600"
        >
          Guest Request
        </label>
        <p
          class="text-gray-900"
        >
          Late check-in requested
        </p>
      </div>
    </div>
    <div
      class="flex justify-end mt-6"
    >
      <div
        data-testid="checkout-modal"
      >
        <div>
          Checkout Modal - 
          COMPLETED
        </div>
        <button>
          Update Status
        </button>
      </div>
    </div>
  </div>
</div>
`;
